### PR TITLE
Boost catchup

### PR DESCRIPTION
**Pull request by codeck**
fixing #1040

### CODE REVIEW

**Comment by jedmccaleb** [comment]
Awesome. before I review. do you have a sense of how much faster this makes it?

**Comment by codeck** [comment]
f846eeb changed the history state json deserialization and bucket name gathering from O(publishqueue length ^ 2)  to O(publishqueue length)
acd36ea changed the complexity of overlap checking from O(# of shared bucket * publishqueue) to O(* # of shared bucket + publishqueue)
161bf82 changed the gc(including bucket name gathering, overlap checking and bucket referencing/dereferencing) times from O(publishqueue) to O(1) , this one makes the former two less signification in performance, but still worth.

the whole process of full catching up was accelerated in about a magnitude in my nodes.

File: src/history/HistoryManagerImpl.cpp [hunk 1]
```diff
@@ -741,4 +770,73 @@ HistoryManagerImpl::getCatchupFailureCount()
...
+        stillEmpty=true;
+        for (auto state :mManager.getPublishQueueStates()) {
+            stillEmpty = false;
+            referenceBucketes(state);
+        }
+    }
+    return stillEmpty;
+}
+
+void HistoryBucketCache::referenceBucketes(HistoryArchiveState &har)
```
Comment (mvaneijk) [inline src/history/HistoryManagerImpl.cpp]: typo: referenceBucket~~e~~s

File: src/history/HistoryManagerImpl.cpp [hunk 2]
```diff
@@ -498,6 +508,8 @@ HistoryManagerImpl::queueCurrentHistory()
     st.define_and_bind();
     st.execute(true);
 
+    getBucketCache().referenceBucketes(has);
```
Comment (mvaneijk) [inline src/history/HistoryManagerImpl.cpp]: typo: referenceBucket~~~e~~~s

**Comment by vogel** [comment]
I've made some measurements.

Change 161bf82 gives speedup of about 3-6 times. Rest of changes does not do additional speedup comparing to just 161bf82 (checked on about 15 000 checkpoints).

As 161bf82 is already merged, I'm closing this PR.

Thank you very much for your contribution!

**Pull request closed without merging**